zkevm-circuits/src/evm_circuit/util/math_gadget/constant_division.rs

```diff
@@ -0,0 +1,163 @@
+        let remainder = numerator % denominator;
+
+        self.quotient
+            .assign(region, offset, Value::known(F::from_u128(quotient)))?;
+        self.remainder
+            .assign(region, offset, Value::known(F::from_u128(remainder)))?;
+
+        self.quotient_range_check
+            .assign(region, offset, F::from_u128(quotient))?;
+
+        Ok((quotient, remainder))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::test_util::*;
+    use super::*;
+    use eth_types::*;
+    use halo2_proofs::halo2curves::bn256::Fr;
+    use halo2_proofs::plonk::Error;
+
+    #[test]
+    fn test_constantdivisiongadget() {
+        const N: usize = 4;
+        #[derive(Clone)]
+        /// a < b
+        struct ConstantDivisionTestContainer<F, const DENOMINATOR: u64, const REMINDER: u64> {
+            constdiv_gadget: ConstantDivisionGadget<F, N>,
+            a: Cell<F>,
+        }
+
+        impl<F: Field, const DENOMINATOR: u64, const REMAINDER: u64> MathGadgetContainer<F>
+            for ConstantDivisionTestContainer<F, DENOMINATOR, REMAINDER>
+        {
+            const NAME: &'static str = "ConstantDivisionGadget";
+
+            fn configure_gadget_container(cb: &mut ConstraintBuilder<F>) -> Self {
+                let a = cb.query_cell();
+                let constdiv_gadget =
+                    ConstantDivisionGadget::<F, N>::construct(cb, a.expr(), DENOMINATOR);
+
+                cb.require_equal(
+                    "a == n * denom",
+                    constdiv_gadget.remainder(),
+                    REMAINDER.expr(),
+                );
+
+                ConstantDivisionTestContainer { constdiv_gadget, a }
+            }
+
+            fn assign_gadget_container(
+                &self,
+                input_words: &[Word],
+                region: &mut CachedRegion<'_, '_, F>,
+            ) -> Result<(), Error> {
+                let a = u64::from_le_bytes(input_words[0].to_le_bytes()[..8].try_into().unwrap());
+                let offset = 0;
+
+                self.a.assign(region, offset, Value::known(F::from(a)))?;
+                self.constdiv_gadget.assign(region, offset, a as u128)?;
+
+                Ok(())
+            }
+        }
+
+        test_math_gadget_container::<Fr, ConstantDivisionTestContainer<Fr, 5, 0>>(
+            vec![Word::from(0)],
+            true,
+        );
+
+        test_math_gadget_container::<Fr, ConstantDivisionTestContainer<Fr, 5, 0>>(
+            vec![Word::from(5)],
+            true,
+        );
+
+        test_math_gadget_container::<Fr, ConstantDivisionTestContainer<Fr, 5, 1>>(
+            vec![Word::from(1)],
+            true,
+        );
+
+        test_math_gadget_container::<Fr, ConstantDivisionTestContainer<Fr, 5, 4>>(
+            vec![Word::from(1)],
+            false,
+        );
+
+        test_math_gadget_container::<Fr, ConstantDivisionTestContainer<Fr, 16, 0>>(
+            vec![Word::from(1 << N)],
+            true,
+        );
+    }
```

Comment:
I think could use some extra test cases that checks that
- remainder < denominator
- Require that quotient < 2**N_BYTES
- The quotient indeed is incorrect (guess this one may be a trickier because needs changing the generated witness)

Reply:
add more cases for these conditions.